general-superstaq/general_superstaq/_models.py

```diff
@@ -0,0 +1,291 @@
+    """A job that involves submitting circuits to live quantum hardware (device)."""
+    DEVICE_SIMULATION = "device_simulation"
+    """A job that involves simulating a circuit as if it was on some live hardware (device).
+    May or may not include noise (depending on options dict)."""
+    SIMULATION = "simulation"
+    """Simulation of a given quantum circuit but agnostic to any specific device."""
+    COMPILE = "compile"
+    """A job that only involves compiling a circuit for some given device."""
+    CONVERT = "convert"
+    """A job that only involves converting circuits between e.g. qiskit and cirq"""
+
+
+class SimMethod(str, Enum):
+    """The different simulation methods that are available."""
+
+    STATEVECTOR = "statevector"
+    """Simulation calculates the wavefunction of the state at the end of the circuit."""
+    NOISE_SIM = "noise-sim"
+    """The simulation applies noise. If used with `device_simulation` this applies realistic device
+    noise otherwise the noise needs to be provided by the user (via the options dict)."""
+    SIM = "sim"
+    """Samples the circuits without any noise."""
+
+
+class CircuitType(str, Enum):
+    """The different languages that are recognized by Superstaq."""
+
+    CIRQ = "cirq"
+    QISKIT = "qiskit"
+    QASM_STRS = "qasm_strs"
+
+
+class CircuitStatus(str, Enum):
+    """The possible statuses of a job/circuit."""
+
+    RECEIVED = "received"
+    """The job has been received (and accepted) to the server and is awaiting further action."""
+    AWAITING_COMPILE = "awaiting_compile"
+    """The job is waiting for a worker to compile."""
+    AWAITING_SUBMISSION = "awaiting_submission"
+    """The job is waiting for a worker to submit the circuit to an external device."""
+    AWAITING_SIMULATION = "awaiting_simulation"
+    """The job is waiting for a worker to simulate."""
+    AWAITING_CONVERSION = "awaiting_conversion"
+    """The job is waiting for a worker to convert."""
+    RUNNING = "running"
+    """The job is being run by a worker."""
+    COMPLETED = "completed"
+    """The job is completed."""
+    FAILED = "failed"
+    """The job failed. A reason should be stored in the job."""
+    CANCELLED = "cancelled"
+    """The job was cancelled. A reason should be stored in the job."""
+    UNRECOGNIZED = "unrecognized"
+    """Something has gone wrong! (Treated as terminal)"""
+    PENDING = "pending"
+    """When a job has been submitted to an external provider but that provider has
+    not yet run the job."""
+    DELETED = "deleted"
+    """When a job has been deleted."""
+
+
+TERMINAL_CIRCUIT_STATES = [
+    CircuitStatus.COMPLETED,
+    CircuitStatus.FAILED,
+    CircuitStatus.CANCELLED,
+    CircuitStatus.UNRECOGNIZED,
+    CircuitStatus.DELETED,
+]
+
+UNSUCCESSFUL_CIRCUIT_STATES = [
+    CircuitStatus.CANCELLED,
+    CircuitStatus.FAILED,
+    CircuitStatus.DELETED,
+]
+
+
+class DefaultPydanticModel(pydantic.BaseModel, use_enum_values=True, extra="forbid"):
```

Comment:
two other options that are nice imo: `validate_assignment=True` and `validate_default=True`